text/0000-swap-out-jemalloc.md

```diff
@@ -0,0 +1,235 @@
+expound on the idea to make it more concrete in terms of what the compiler
+implementation will look like. A [sample implementaiton][impl] is available of
+this section.
+
+[gist]: https://gist.github.com/alexcrichton/41c6aad500e56f49abda
+[impl]: https://github.com/alexcrichton/rust/tree/less-jemalloc
+
+### High level design
+
+The design of this RFC from 10,000 feet (referred to below), which was
+[previously outlined][gist] looks like:
+
+1. Define a set of symbols which correspond to the APIs specified in
+   `alloc::heap`. The `liballoc` library will call these symbols directly.
+   Note that this means that each of the symbols take information like the size
+   of allocations and such.
+2. Create two shim libraries which implement these allocation-related functions.
+   Each shim is shipped with the compiler in the form of a static library. One
+   shim will redirect to the system allocator, the other shim will bundle a
+   jemalloc build along with Rust shims to redirect to jemalloc.
+3. Intermediate artifacts (rlibs) do not resolve this dependency, they're just
+   left dangling.
+4. When producing a "final artifact", rustc by default links in one of two
+   shims:
+    * If we're producing a staticlib or a dylib, link the system shim.
+    * If we're producing an exe and all dependencies are rlibs link the
+      jemalloc shim.
+
+The final link step will be optional, and one could link in any compliant
+allocator at that time if so desired.
+
+### New Attributes
+
+Two new **unstable** attributes will be added to the compiler:
+
+* `#![needs_allocator]` indicates that a library requires the "allocation
+  symbols" to link successfully. This attribute will be attached to `liballoc`
+  and no other library should need to be tagged as such. Additionally, most
+  crates don't need to worry about this attribute as they'll transitively link
+  to liballoc.
+* `#![allocator]` indicates that a crate is an allocator crate. This is
+  currently also used for tagging FFI functions as an "allocation function"
+  to leverage more LLVM optimizations as well.
+
+All crates implementing the Rust allocation API must be tagged with
+`#![allocator]` to get properly recognized and handled.
+
+### New Crates
+
+Two new **unstable** crates will be added to the standard distribution:
+
+* `alloc_system` is a crate that will be tagged with `#![allocator]` and will
+  redirect allocation requests to the system allocator.
+* `alloc_jemalloc` is another allocator crate that will bundle a static copy of
+  jemalloc to redirect allocations to.
+
+Both crates will be available to link to manually, but they will not be
+available in stable Rust to start out.
+
+### Allocation functions
+
+Each crate tagged `#![allocator]` is expected to provide the full suite of
+allocation functions used by Rust, defined as:
+
+```rust
+extern {
+    fn __rust_allocate(size: usize, align: usize) -> *mut u8;
```

Comment:
Implementation-wise, this is what everything will boil down to (pre-defined symbols), and this is currently the path of least resistance forward. This is all unstable, however, so we'll definitely be able to change it in the future to perhaps using lang items or more official attributes. The current downside of attributes are:

* During a compilation, there may actually be two loaded allocators in the crate store (but we won't link one of them), so the compiler would detect duplicate lang items and yield an error. Extra logic would have to be added to "not worry about" the allocator lang items.
* None of the signatures are currently typechecked, and having an official attribute makes it feel like it *should* be typechecked.

Basically I'd love to move to using attributes and such, but I don't see much immediate benefit over just defining some symbols in the short-term. I also don't mind adding some words to this effect in the RFC, though, and we could perhaps spec the "ideal implementation" here where the actual implementation just has some TODOs.

My ideal situation would be to have an attribute-per-function which defines the symbol, visibility, and typechecks the signature. We'd then also have a check that an `#![allocator]` crate contains the necessary functions (tagged with attributes). That's a good deal of attribute-surface-area to start stabilizing right off the bat though.